pkg/controller/direct/spanner/instance_controller.go

```diff
@@ -0,0 +1,284 @@
+	id         *krm.SpannerInstanceRef
+	gcpClient  *gcp.InstanceAdminClient
+	desired    *krm.SpannerInstance
+	actual     *spannerpb.Instance
+	resourceID string
+}
+
+var _ directbase.Adapter = &SpannerInstanceAdapter{}
+
+func (a *SpannerInstanceAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("getting SpannerInstance", "name", a.id.External)
+
+	req := &spannerpb.GetInstanceRequest{Name: a.id.External}
+	instancepb, err := a.gcpClient.GetInstance(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting SpannerInstance %q: %w", a.id.External, err)
+	}
+
+	a.actual = instancepb
+	return true, nil
+}
+
+func (a *SpannerInstanceAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("creating Instance", "name", a.id.External)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	if err := a.SpecValidation(); err != nil {
+		return err
+	}
+	resource := SpannerInstanceSpec_ToProto(mapCtx, &desired.Spec, a.id.SpannerInstanceConfigPrefix())
+	// If node count or processing unit is not specify,
+	// Default NodeCount to 1.
+	if resource.NodeCount == 0 && resource.ProcessingUnits == 0 {
+		resource.NodeCount = 1
+	}
+	resource.Name = a.id.External
+	resource.Labels = desired.Labels
+	resource.Labels["managed-by-cnrm"] = "true"
+	resource.Name = a.id.External
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	parent, err := a.id.Parent()
+	if err != nil {
+		return err
+	}
+
+	req := &spannerpb.CreateInstanceRequest{
+		InstanceId: a.resourceID,
+		Instance:   resource,
+		Parent:     parent.String(),
+	}
+	op, err := a.gcpClient.CreateInstance(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating Instance %s: %w", a.id.External, err)
+	}
+	created, err := op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("Instance %s waiting creation: %w", a.id.External, err)
+	}
+	log.V(2).Info("successfully created Instance", "name", a.id.External)
+
+	status := &krm.SpannerInstanceStatus{}
+	status.State = State_FromProto(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ExternalRef = &a.id.External
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+func (a *SpannerInstanceAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("updating Instance", "name", a.id.External)
+	mapCtx := &direct.MapContext{}
+	if err := a.SpecValidation(); err != nil {
+		return err
+	}
+	desired := a.desired.DeepCopy()
+	resource := SpannerInstanceSpec_ToProto(mapCtx, &desired.Spec, a.id.SpannerInstanceConfigPrefix())
+	resource.Name = a.id.External
+	resource.Labels = desired.Labels
+	resource.Labels["managed-by-cnrm"] = "true"
+	resource.Name = a.id.External
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	updateMask := &fieldmaskpb.FieldMask{}
```

Comment:
Suggest using the `common.CompareProtoMessage` here?

Reply:
BTW, we can improve the CompareProtoMessage to apply comparison with customized conditions.